File: moseoh/_11주차/쿼드압축_후_개수_세기/java/Solution.java
package _11주차.쿼드압축_후_개수_세기.java;

import java.util.Arrays;

class Solution {
    public final static int ALREADY_COMPRESS = -1;

    public static void main(String[] args) {
        int[][] arr1 = new int[][]{{1, 1, 0, 0}, {1, 0, 0, 0}, {1, 0, 0, 1}, {1, 1, 1, 1}};
        int[] answer1 = new int[]{4, 9};
        int[] result1 = new Solution().solution(arr1);
        PRINT_RESULT(1, result1, answer1);

        int[][] arr2 = new int[][]{{1, 1, 1, 1, 1, 1, 1, 1}, {0, 1, 1, 1, 1, 1, 1, 1}, {0, 0, 0, 0, 1, 1, 1, 1}, {0, 1, 0, 0, 1, 1, 1, 1}, {0, 0, 0, 0, 0, 0, 1, 1}, {0, 0, 0, 0, 0, 0, 0, 1}, {0, 0, 0, 0, 1, 0, 0, 1}, {0, 0, 0, 0, 1, 1, 1, 1}};
        int[] answer2 = new int[]{10, 15};
        int[] result2 = new Solution().solution(arr2);
        PRINT_RESULT(2, result2, answer2);
    }

    public static void PRINT_RESULT(int index, int[] result, int[] answer) {
        boolean correct = Arrays.equals(result, answer);
        StringBuilder sb = new StringBuilder();
        sb.append("\n\n테스트 케이스 ").append(index).append(": ");
        sb.append(correct ? "정답" : "오답").append("\n");
        sb.append("\t- 실행 결과: \t").append(Arrays.toString(result)).append("\n");
        sb.append("\t- 기댓값: \t").append(Arrays.toString(answer)).append("\n");
        if (correct) System.out.println(sb);
        else throw new RuntimeException(sb.toString());
    }

    public int[] solution(int[][] arr) {
        int[] answer = {0, 0};
        int n = arr.length;

        for (int i = n; i > 0; i /= 2) {
            for (int x = 0; x < n; x += i) {
                for (int y = 0; y < n; y += i) {
                    if (canCompress(arr, x, y, i)) {
                        answer[arr[x][y]]++;
                        compress(arr, x, y, i);
                    }
                }
            }
        }

        return answer;
    }

    public boolean canCompress(int[][] arr, int indexX, int indexY, int length) {
        int target = arr[indexX][indexY];
        if (target == ALREADY_COMPRESS) return false;

        for (int x = indexX; x < indexX + length; x++) {
            for (int y = indexY; y < indexY + length; y++) {
                if (arr[x][y] != target) return false;
            }
        }
        return true;
    }

    public void compress(int[][] arr, int indexX, int indexY, int length) {
        for (int x = indexX; x < indexX + length; x++) {
            for (int y = indexY; y < indexY + length; y++) {
                arr[x][y] = ALREADY_COMPRESS;
            }
        }
    }
}

File: moseoh/_11주차/쿼드압축_후_개수_세기/kotlin/Solution.kt
package _11주차.쿼드압축_후_개수_세기.kotlin

fun main() {
    fun printResult(index: Int, result: IntArray, answer: IntArray) {
        val correct = result.contentEquals(answer)
        val sb = StringBuilder()
        sb.append("\n\n테스트 케이스 ").append(index).append(": ")
        sb.append(if (correct) "정답" else "오답").append("\n")
        sb.append("\t- 실행 결과: \t").append(result.contentToString()).append("\n")
        sb.append("\t- 기댓값: \t").append(answer.contentToString()).append("\n")
        if (correct) println(sb) else throw RuntimeException(sb.toString())
    }

    val arr1 = arrayOf(intArrayOf(1, 1, 0, 0), intArrayOf(1, 0, 0, 0), intArrayOf(1, 0, 0, 1), intArrayOf(1, 1, 1, 1))
    val answer1 = intArrayOf(4, 9)
    val result1 = Solution().solution(arr1)
    printResult(1, result1, answer1)

    val arr2 = arrayOf(
        intArrayOf(1, 1, 1, 1, 1, 1, 1, 1),
        intArrayOf(0, 1, 1, 1, 1, 1, 1, 1),
        intArrayOf(0, 0, 0, 0, 1, 1, 1, 1),
        intArrayOf(0, 1, 0, 0, 1, 1, 1, 1),
        intArrayOf(0, 0, 0, 0, 0, 0, 1, 1),
        intArrayOf(0, 0, 0, 0, 0, 0, 0, 1),
        intArrayOf(0, 0, 0, 0, 1, 0, 0, 1),
        intArrayOf(0, 0, 0, 0, 1, 1, 1, 1)
    )
    val answer2 = intArrayOf(10, 15)
    val result2 = Solution().solution(arr2)
    printResult(2, result2, answer2)
}

class Solution {
    fun solution(arr: Array<IntArray>): IntArray {
        val answer = intArrayOf(0, 0)
        val n = arr.size
        for (i in generateSequence(n) { it / 2 }.takeWhile { it > 0 }) {
            for (x in 0 until n step i) {
                for (y in 0 until n step i) {
                    if (canCompress(arr, x, y, i)) {
                        answer[arr[x][y]]++
                        compress(arr, x, y, i)
                    }
                }
            }
        }
        return answer
    }

    private fun canCompress(arr: Array<IntArray>, indexX: Int, indexY: Int, length: Int): Boolean {
        val target = arr[indexX][indexY]
        if (target == -1) return false
        for (x in indexX until indexX + length) {
            for (y in indexY until indexY + length) {
                if (arr[x][y] != target) return false
            }
        }
        return true
    }

    private fun compress(arr: Array<IntArray>, indexX: Int, indexY: Int, length: Int) {
        for (x in indexX until indexX + length) {
            for (y in indexY until indexY + length) {
                arr[x][y] = -1
            }
        }
    }
}